File: leetcode/123-best-time-to-buy-and-sell-stock-iii.cpp
/*
Say you have an array for which the ith element is the price of a given stock on day i.

Design an algorithm to find the maximum profit. You may complete at most two transactions.

Note:
You may not engage in multiple transactions at the same time (ie, you must sell the stock before you buy again).
*/

using sol_t = tuple<int, int, int>;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        vector<vector<int>> states { {INT_MIN, 0}, {0, 0}, {INT_MIN, 0}, {0, 0}};
        int current = 0, next = 1;
        for (auto price : prices) {
            states[0][next] = max(states[0][current], -price);
            states[1][next] = max(states[1][current], states[0][current] + price);
            states[2][next] = max(states[2][current], states[1][current] - price);
            states[3][next] = max(states[3][current], states[2][current] + price);
            swap(current, next);
        }
        return max(states[1][current], states[3][current]);
    }
};

File: leetcode/123-best-time-to-buy-and-sell-stock-iii.rb
# @param {Integer[]} prices
# @return {Integer}
def max_profit(prices)
  #    [current, next]
  s1 = [nil, 0] # open trans1
  s2 = [0, 0]   # close trans1
  s3 = [nil, 0] # open trans2
  s4 = [0, 0]   # close trans2
  current = 0
  nxt = 1
  prices.each do |price|
    s1[nxt] = s1[current].nil? ? -price : [s1[current], -price].max
    s2[nxt] = s1[current].nil? ? 0 : [s2[current], s1[current] + price].max
    s3[nxt] = s3[current].nil? ? -price : [s3[current], s2[current] - price].max
    s4[nxt] = s3[current].nil? ? 0 : [s4[current], s3[current] + price].max
    nxt, current = current, nxt
  end
  [s2[current], s4[current]].max
end

